cmd/thanos/tools_bucket.go

```diff
@@ -438,6 +440,11 @@ func registerBucketReplicate(m map[string]setupFunc, root *kingpin.CmdClause, na
 	compactions := cmd.Flag("compaction", "Only blocks with these compaction levels will be replicated. Repeated flag.").Default("1", "2", "3", "4").Ints()
 	matcherStrs := cmd.Flag("matcher", "Only blocks whose external labels exactly match this matcher will be replicated.").PlaceHolder("key=\"value\"").Strings()
 	singleRun := cmd.Flag("single-run", "Run replication only one time, then exit.").Default("false").Bool()
+	minTime := model.TimeOrDuration(cmd.Flag("min-time", "Start of time range limit to replicate. Thanos Replicate will replicate only metrics, which happened later than this value. Option can be a constant time in RFC3339 format or time duration relative to current time, such as -1d or 2h45m. Valid duration units are ms, s, m, h, d, w, y.").
+		Default("0000-01-01T00:00:00Z"))
+	maxTime := model.TimeOrDuration(cmd.Flag("max-time", "End of time range limit to replicate. Thanos Replicate will replicate only metrics, which happened earlier than this value. Option can be a constant time in RFC3339 format or time duration relative to current time, such as -1d or 2h45m. Valid duration units are ms, s, m, h, d, w, y.").
+		Default("9999-12-31T23:59:59Z"))
+	deleteOldBlocks := cmd.Flag("delete-old-blocks", "Delete blocks that are older then max-time.").Default("false").Bool()
```

Comment:
I don't think replicate should ever be deleting data

Reply:
Eh this should be min-time actually and not max time 🤦 . 

But the idea here is that, replicate will only replicate data that is inside the window of --max and --min time, e.g. -10h to -30d, to ensure that the bucket only has the data inside this window, there must be a flag to delete the data that falls out of this window.

Reply:
Yes, let's use compactor for that.

Reply:
@bwplotka how if I may ask? Should replicator still mark the blocks for deletion? In our particular case, we were not planning on running compactor on the bucket.

Reply:
Plus its an optional flag 😔

Reply:
Just to be clear, this flag is suppose to delete the old blocks from the `toBkt` to maintain the retention set on the min-max flags. 

So your suggestion is to add a functionality to compactor that goes over all the blocks and deletes the ones that are outside of the retention? I thought the whole point of compactor is to compact data. To me it makes more sense to add this to replicator as you want to replicate and have data in a certain time window. 

So to me it makes more sense to have replicator mark blocks for deletion and then let compactor or maybe even a whole new tool dedicated to clean up, delete these blocks. 🤔 And then compactor also marks things for deletion and then this cleanup tool ends up deleting them.

Reply:
Because in the end, we don't need compactor to compact data on the `toBkt`.

Reply:
Ok I see that compactor can run without downsampling and actually just delete old data. So I guess compactor can do the job.

Reply:
After testing this setup with compactor, I think this will not work. Im running into the overlap issue as explained here: https://github.com/thanos-io/thanos/blob/master/docs/operating/troubleshooting.md#overlaps.

And the only explanation I have is because we are no running 2 compactors on the same set of blocks. 

We are replicating from bucket 1 to bucket 2, bucket 1 has compactor running on it originally. We were not planning on running compactor on bucket 2 because its already running on bucket 1. But because of your suggestion to use compactor with downsampling disabled and only use it for retention, it still halts when trying to compact blocks due to overlap.


I still think that

>So to me it makes more sense to have replicator mark blocks for deletion and then let compactor or maybe even a whole new tool dedicated to clean up, delete these blocks. 🤔 And then compactor also marks things for deletion and then this cleanup tool ends up deleting them.

is the way to go. Move the deletion of marked blocks to another tool and let compactor and replicator mark blocks for deletion. WDYT?

Reply:
Let's talk offline (: